torch/csrc/jit/script/compiler.cpp

```diff
@@ -955,6 +956,49 @@ struct to_ir {
     return emitIfExpr(expr.range(), cond_value, true_expr, false_expr);
   }
 
+  Value* emitListComprehension(const ListComp& lc) {
+    // this avoids a race condition where we would re-use the same temp name
+    static std::atomic<size_t> tmp_count{0};
+    auto tmp_name = std::string("___list_acc") + std::to_string(tmp_count++);
+    auto list_value = emitExpr(lc.iter());
+    if (list_value->type()->kind() != TypeKind::ListType) {
+      // TODO: constraining iterators to be simple lists for now
+      // as it makes easy to get list's element type.
+      throw ErrorReport(lc.range())
+          << "iterator expression is expected to be a list";
+    }
+    auto elem_types = list_value->type()->containedTypes();
+    // TODO: users can easily change the type to (x,1) or float(x)
+    // as in `float(x) for x in my_list_of_ints`
+    // eventually, we would probably want to temporarily inject x
+    // so we can evaluate the generator expression (e.g. `float(x)`) depending
+    // on x
+
+    // given `[x*2 for x in my_list]` this generates the following AST:
+    // __list_acc = []
+    // for x in my_list:
+    //  __list_acc.append(x*2)
+    auto n = graph->insertNode(
+        graph->createList(elem_types.at(0), at::ArrayRef<Value*>{}));
+    environment_stack->setVar(lc.range(), tmp_name, n->output());
+    auto tmp_list_ident = Ident::create(lc.range(), tmp_name);
+    auto tmp_list_var = Var::create(lc.range(), tmp_list_ident);
+    auto append_ident = Ident::create(lc.range(), "append");
+    auto dot_op = Select::create(lc.range(), tmp_list_var, append_ident);
+    auto append_args_list = List<Expr>::create(lc.range(), {lc.elt()});
+    auto append_attrs = List<Attribute>::create(lc.range(), {});
+    auto apply_append =
+        Apply::create(lc.range(), dot_op, append_args_list, append_attrs);
+    auto expr_stmt = ExprStmt::create(lc.range(), apply_append);
+    auto stmt_list = List<Stmt>::create(lc.range(), {expr_stmt});
+    auto iters_list = List<Expr>::create(lc.range(), {lc.iter()});
+    auto targets_list = List<Expr>::create(lc.range(), {lc.target()});
+    auto for_loop =
+        For::create(lc.range(), targets_list, iters_list, stmt_list);
+    emitFor(for_loop);
+    return n->output();
```

Comment:
Return the output of the prim::Loop node here instead of the original ListConstruct

Reply:
This comment is still valid I believe

Reply:
`emitFor` doesn't return a `Value` and `%8` isn't automatically added to loop outputs. Neither one is insurmountable but it will add even **more** boilerplate.

```
graph(%l : int[]):
  %1 : bool = prim::Constant[value=1]()
  %2 : int = prim::Constant[value=3]()
  %___list_acc0 : int[] = prim::ListConstruct()
  %4 : int = aten::len(%l)
   = prim::Loop(%4, %1)
    block0(%5 : int):
      %x : int = aten::select(%l, %5)
      %7 : int = aten::mul(%x, %2)
      %8 : int[] = aten::append(%___list_acc0, %7)
      -> (%1)
  return (%___list_acc0)
```